pkg/test/parameter.go

```diff
@@ -7,23 +7,52 @@ package test
 
 import (
 	"bytes"
+	"encoding/json"
 	"errors"
 	"fmt"
 	"text/template"
 
 	"github.com/facebookincubator/contest/pkg/target"
 )
 
-// Param represents a test step parameter. It is initialized from a string,
-// which can be either a host name, a function expression, or just any regular
-// string. It enables hostname and function expression validation and a few
-// other validation operations. This is meant to simplify the way plugin access
-// and use test step parameters.
-type Param string
+// NewParam inititalizes a new Param object directly from a string.
+// Note that no validation is performed if the input is actually valid JSON.
+// This helper is primarily intended for tests.
+func NewParam(s string) *Param {
+	var p Param
+	p.RawMessage = json.RawMessage(s)
+	return &p
+}
+
+// Param represents a test step parameter. It is initialized from JSON,
+// and can be a string or a more complex JSON structure.
+// Plugins are expected to know which one they expect and use the
+// provided convenience functions to obtain either the string or
+// json.RawMessage representation.
+type Param struct {
+	json.RawMessage
+}
 
 // IsEmpty returns true if the original raw string is empty, false otherwise.
 func (p Param) IsEmpty() bool {
-	return p == ""
+	return p.String() == ""
+}
+
+// String returns the parameter as a string. This helper never fails.
+// If the underlying JSON cannot be unmarshalled into a simple string,
+// this function returns a string representation of the JSON structure.
+func (p Param) String() string {
+	var str string
+	if json.Unmarshal(p.RawMessage, &str) == nil {
+		return str
+	}
+	// can't unmarshal to string, return raw json
+	return string(p.RawMessage)
+}
+// JSON returns the parameter as json.RawMessage for further
+// unmarshalling by the test plugin.
+func (p Param) JSON() json.RawMessage {
+	return p.RawMessage
```

Comment:
should `RawMessage` also be private? With the helpers it seems we don't need direct access to it

Reply:
Oh, can you control the visibility of this kind of struct embedding somehow?

Reply:
I am reading the interwebs and it seems I can't make the embedded field private :/

Reply:
not the embedded field itself, but if we have a `JSON` method this could become a non-anonymous field. However this is good as it is, I was just wondering if we could handle that differently, I'll merge now